grouper/permissions.py

```diff
@@ -0,0 +1,245 @@
+    """No owner was found for the permission + argument combination."""
+
+
+class InvalidRequestID(PermissionRequestException):
+    """Submitted request ID is invalid (doesn't exist or group doesn't match."""
+
+
+def create_request(session, user, group, permission, argument, reason):
+    """
+    Creates an permission request and sends notification to the responsible approvers.
+
+    Args:
+        session(sqlalchemy.orm.session.Session): database session
+        user(models.User): user requesting permission
+        group(models.Group): group requested permission would be applied to
+        permission(models.Permission): permission in question to request
+        argument(str): argument for the given permission
+        reason(str): reason the permission should be granted
+
+    Raises:
+        RequestAlreadyExists if trying to create a request that is already pending
+        NoOwnersAvailable if no owner is available for the requested perm + arg.
+    """
+    # check if request already pending for this perm + arg pair
+    existing_count = session.query(PermissionRequest).filter(
+            PermissionRequest.group_id == group.id,
+            PermissionRequest.permission_id == permission.id,
+            PermissionRequest.argument == argument,
+            PermissionRequest.status == "pending",
+            ).count()
+
+    if existing_count > 0:
+        raise RequestAlreadyExists()
+
+    # determine owner(s)
+    owners = get_owners(session, permission, argument)
+
+    if not owners:
+        raise NoOwnersAvailable()
+
+    pending_status = "pending"
+    now = datetime.utcnow()
+
+    # multiple steps to create the request
+    request = PermissionRequest(
+            requester_id=user.id,
+            group_id=group.id,
+            permission_id=permission.id,
+            argument=argument,
+            status=pending_status,
+            requested_at=now,
+            ).add(session)
+    session.flush()
+
+    request_status_change = PermissionRequestStatusChange(
+            request=request,
+            user=user,
+            to_status=pending_status,
+            change_at=now,
+            ).add(session)
+    session.flush()
+
+    Comment(
+            obj_type=OBJ_TYPES_IDX.index("PermissionRequestStatusChange"),
+            obj_pk=request_status_change.id,
+            user_id=user.id,
+            comment=reason,
+            created_on=now,
+            ).add(session)
+
+    # send notification
+    email_context = {
+            "user_name": user.name,
+            "group_name": group.name,
+            "permission_name": permission.name,
+            "argument": argument,
+            "reason": reason,
+            }
+
+    # TODO: would be nicer if it told you which group you're an approver of
+    # that's causing this notification
+    mail_to = []
+    for owner in owners:
+        mail_to += [u.name for u in owner.my_users() if u.role in APPROVER_ROLE_INDICIES]
```

Comment:
I think you want all the group members here, not just the ones in `APPROVER_ROLE_INDICIES`.  My mental model is that the Grouper permission to grant permissions applies to the entire group and any member can do the grant, so the email notification should go to all the members of the group.  That will be noisy for groups with lots of members, but I think groups with lots of members generally shouldn't have the ability to grant other permissions anyway.